### PR TITLE
Julient/schema coders

### PR DESCRIPTION
**Pull request by jto**
**WIP -- DO NOT MERGE**

This branch adds support for Schema Coders and BeamSQL

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala [hunk 1]
```diff
@@ -131,18 +134,33 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
...
     context.wrap(internal.setCoder(coder))
 
+  def setSchema(schema: Schema[T]): SCollection[T] = {
+    if (!internal.hasSchema) {
+      val (s, to, from) = SchemaMaterializer.materialize(this.context, schema)
+      context.wrap(internal.setSchema(s, to, from))
+    } else this
+  }
+
+  private def checkCoderSerialization[T](coder: BCoder[T]): Unit =
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala]: wdyt of returning `Either` here? I know that we have this style across the code base but I would like to start changing this behavior

File: scio-macros/src/main/scala/com/spotify/scio/IsJava.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Proof that a type is implemented in Java
+ */
+sealed trait IsJavaBean[T]
+
+object IsJavaBean {
+  implicit def isJavaBean[T]: IsJavaBean[T] = macro IsJavaBean.isJavaBeanImpl[T]
+
+  def apply[T](implicit i: IsJavaBean[T]): IsJavaBean[T] = i
```
Comment (regadas) [inline scio-macros/src/main/scala/com/spotify/scio/IsJava.scala]:
```suggestion
  @inline final def apply[T](implicit i: IsJavaBean[T]): IsJavaBean[T] = i
```

File: scio-core/src/main/scala/com/spotify/scio/sql/Query.scala [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+      val query = q
+      def apply(c: SCollection[I]) = {
+
+        // XXX: Hack to set the coder on the existing PCOLLECTION
+        // and hide it in a "set schema" dataflow block
+        val coder = {
+          val (schema, to, from) = SchemaMaterializer.materialize(c.context, Schema[I])
+          Coder.beam(SchemaCoder.of(schema, to, from))
+        }
+        val scoll = c.transform("set schema")(_.map(identity)(coder)).setSchema(Schema[I])
```
[the remark this answers is not in the document]
Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/sql/Query.scala]: I get it now, the coder might have been set already and we can't change it 👍

File: scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala (hunk 1 above)
Comment (jto) [reply scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala]: This is somewhat out of the scope of this PR. I'd prefer to make that change in another one

Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala]: +1

File: scio-core/src/main/scala/com/spotify/scio/schemas/To.scala [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        }
+    }
+  }
+
+  private[schemas] def checkCompatibility[T](bsi: BSchema, bso: BSchema)(
+    t: => T): Either[String, T] = {
+    if (areCompatible(bsi, bso)) {
+      Right(t)
+    } else {
+      val message =
```
[the remark this answers is not in the document]
Comment (jto) [reply scio-core/src/main/scala/com/spotify/scio/schemas/To.scala]: Not really. It uses `bsi` and `bso`

File: scio-core/src/main/scala/com/spotify/scio/schemas/To.scala [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      case (BSchema.TypeName.MAP, BSchema.TypeName.MAP, _) =>
+        areCompatible(t0.getMapKeyType, t1.getMapKeyType)
+        areCompatible(t0.getMapValueType, t1.getMapValueType)
+      case (_, _, _) =>
+        t0.equivalent(t1, BSchema.EquivalenceNullablePolicy.SAME)
+    }
+  }
+
+  private def areCompatible(s0: BSchema, s1: BSchema): Boolean = {
+    val s0Fields = s0.getFields.asScala
```
[the remark this answers is not in the document]
Comment (jto) [reply scio-core/src/main/scala/com/spotify/scio/schemas/To.scala]: Yep. Good catch!

**Pull request merged**